### PR TITLE
Split storage and compute within dataflow

### PR DESCRIPTION
**Pull request by frankmcsherry**
This PR creates `storage` and `compute` modules in the `mz-dataflow` crate, for the purposes of partitioning the logic that should be associated with each of these layers. The work is only just starting, and this PR is exclusively code movement rather than any behavioral change.

Still to do:
1. Extract `{Storage, Compute}State` into their modules.
2. Consolidate `boundary` and `sink` functionality into a `StorageClient` under `storage` but intended for consumption by `compute`.

### Motivation

   * This PR refactors existing code.

We would like to partition the implementations of the storage and compute layers, to allow for independent development and evolution. The two are tangled at the moment, and the in-crate partitioning is meant to allow us to pre-arrange the files and modules so that they may be more cleanly extracted.

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Review by petrosagg — COMMENTED** [review]
everything looks reasonable! It's unclear from the description if you intend to do more work on this before merging or if this first batch of work should land. I'm happy with this landing as-is, I'll click the approve button depending on the answer

**Comment by frankmcsherry** [comment]
I'm thinking I might land this, as I am afraid of future conflicts on this stuff most (and less on the state and boundaries).

File: src/dataflow/src/logging/timely.rs [hunk 1]
```diff
@@ -22,8 +22,9 @@ use timely::logging::{ParkEvent, TimelyEvent, WorkerIdentifier};
 
 use super::{LogVariant, TimelyLog};
 use crate::activator::RcActivator;
-use crate::arrangement::manager::RowSpine;
-use crate::arrangement::KeysValsHandle;
+use mz_dataflow_types::KeysValsHandle;
+use mz_dataflow_types::RowSpine;
+
```
Comment (guswynn) [inline src/dataflow/src/logging/timely.rs:27]: The way these 2 sections are laid out is a bit weird now (as well as the other changes like this one), overall to match the rest of the repo it should probably be

```
use std::whatever;
use std::whatever2;

use external_crate::a;
use external_crate2::b;

use mz_crate::stuff;
use mz_crate2::b;

use super::stuff;
use crate::crate_stuff;
```

Comment (guswynn) [reply src/dataflow/src/logging/timely.rs:27]: unfortunately rustfmt only sorts within each block separated by spaces, and the options that make it split imports into sections are nightly :(

File: src/dataflow/src/storage/render/mod.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Renders a plan into a timely/differential dataflow computation.
```
Comment (guswynn) [inline src/dataflow/src/storage/render/mod.rs:10]: do these docs make more sense in `compute::render`?

**Review by guswynn — APPROVED** [review]
looks good minus some smaller comments

curious, why would be put `sink` and `boundary` in `storage`, not `compute`?

**Comment by frankmcsherry** [comment]
> curious, why would be put sink and boundary in storage, not compute?

They are meant to be defined by the Storage team, even though they run on a compute instance. E.g. boundary controls "how do we get the definite data behind a source" and the storage team is the one that gets to decide "let's switch that to pull from S3 rather than use TCP to a local storage instance".

**Comment by guswynn** [comment]
@frankmcsherry makes sense!

File: src/dataflow/src/logging/timely.rs (hunk 1 above)
Comment (frankmcsherry) [reply src/dataflow/src/logging/timely.rs:27]: I just type `cargo fmt`. Happy to have folks with additional style requirements come in afterwards!

**Review by antiguru — APPROVED** [review]

**Comment by frankmcsherry** [comment]
I'll track down some of the importing stuff in the next PR, @guswynn, but my general sense is that there are many other issues with the code quality, and generally grabbing some time for all of us to tidy things up would be even better.

**Pull request merged**